File: quant1x/contrib/data/tdx/bar_raw.cpp
#include <quant1x/contrib/data/tdx/bar_raw.h>
#include <quant1x/contrib/data/tdx/client.h>
#include <quant1x/contrib/data/tdx/instruments.h>
#include <quant1x/contrib/data/tdx/level1/std/security_bars.h>
#include <quant1x/contrib/data/tdx/level1/ext/instrument_bars.h>
#include <quant1x/config/base.h>
#include <quant1x/data/base.h>
#include <quant1x/data/meta/exchange.h>
#include <quant1x/io/csv-reader.h>
#include <spdlog/spdlog.h>
#include <fmt/format.h>
#include <quant1x/io/csv-writer.h>
#include <filesystem>
#include <algorithm>
#include <cctype>

namespace config = quant1x::config;
namespace data = quant1x::data;
namespace csvio = ::io;
namespace meta = quant1x::data;
using quant1x::contrib::data::tdx::KLineType;
using quant1x::contrib::data::tdx::SecurityBarsContext;

namespace quant1x::contrib::data::tdx {

// =============================
// 常量 (对齐 Python kline_raw.py)
// =============================

/// 每页请求的最大K线数量 (对齐 Python SECURITY_BARS_PRE_REQUEST_MAX)
constexpr int kSecurityBarsPreRequestMax = 800;

/// 增量更新缓存清理的最大天数 (对齐 Python MaxCachedDaysToDropOnIncrementalUpdate)
constexpr int kMaxCachedDaysToDrop = 1;

/// 全局默认起始日期 (对齐 Python GLOBAL_DEFAULT_START_DATE)
constexpr const char* kGlobalDefaultStartDate = "1990-12-19";

// =============================
// BarRaw — 缓存格式 (对齐 Python kline_raw.py BarRaw)
// =============================

struct BarRaw {
    std::string date;
    double      open = 0.0;
    double      close = 0.0;
    double      high = 0.0;
    double      low = 0.0;
    double      volume = 0.0;
    double      amount = 0.0;
    int         up = 0;
    int         down = 0;
    std::string timestamp;

    /// 从 domain Bar 构造 (对齐 Python: for row in vec: BarRaw(date=..., open=row.open, ...))
    static BarRaw from_bar(const schema::Bar& bar) {
        return BarRaw{
            bar.date,
            bar.open,
            bar.close,
            bar.high,
            bar.low,
            bar.volume,
            bar.amount,
            bar.up,
            bar.down,
            bar.timestamp
        };
    }
};

// =============================
// Raw K线缓存 I/O — BarRaw 格式 (对齐 Python kline_raw.py save_kline_raw / read_kline_raw_from_csv)
// =============================

static std::string get_kline_raw_filename(const meta::Instrument& inst) {
    return config::default_cache_path() + "/day_raw/" + inst.cache_dir() + "/" + inst.symbol() + ".raw";
}

static void save_kline_raw(const std::string& filename, const std::vector<BarRaw>& values) {
    if (values.empty()) return;
    auto dir = std::filesystem::path(filename).parent_path().string();
    std::filesystem::create_directories(dir);

    csvio::CSVWriter writer(filename);
    writer.write_row("date", "open", "close", "high", "low", "volume", "amount", "up", "down", "timestamp");
    for (const auto& v : values) {
        writer.write_row(v.date, v.open, v.close, v.high, v.low,
                         v.volume, v.amount, v.up, v.down, v.timestamp);
    }
}

static std::vector<BarRaw> read_kline_raw_from_csv(const std::string& filename) {
    std::vector<BarRaw> klines;
    try {
        csvio::CSVReader<10> in(filename);
        in.read_header(csvio::ignore_extra_column, "date", "open", "close", "high", "low",
                       "volume", "amount", "up", "down", "timestamp");
        BarRaw row = {};
        while (in.read_row(row.date, row.open, row.close, row.high, row.low,
                           row.volume, row.amount, row.up, row.down, row.timestamp)) {
            klines.emplace_back(std::move(row));
        }
    } catch (const std::exception& e) {
        spdlog::warn("[kline_raw] read_kline_raw_from_csv error: {}", e.what());
    }
    return klines;
}

// =============================
// fetch_kline_raw (对应 Python fetch_kline_raw — 返回 domain schema Bar)
// =============================

/// fetch_kline_raw_from_std — 从标准行情获取原始K线, 转换为 domain Bar
/// 对应 Python kline_raw.py fetch_kline_raw_from_std:
///   msg = SecurityBarsContext(inst.exchange, inst.ticker, kline_type, start, count, inst.type.is_index())
///   protocol.transact_message_sync(conn, msg)
///   return msg.list  # msg.list 已经是 List[Bar]
static std::vector<schema::Bar> fetch_kline_raw_from_std(
        const meta::Instrument& inst, int start, int count, u16 category) {
    try {
        auto conn = get_std_conn();
        SecurityBarsContext bars(inst, category,
                                  static_cast<u16>(start), static_cast<u16>(count));
        transact_message_sync(conn->socket(), bars);

        std::vector<schema::Bar> result;
        result.reserve(bars.List.size());
        for (auto const& raw : bars.List) {
            schema::Bar bar;
            bar.date             = raw.DateTime.substr(0, 10);
            bar.open             = raw.Open;
            bar.close            = raw.Close;
            bar.high             = raw.High;
            bar.low              = raw.Low;
            bar.volume           = raw.Vol * 100;    // 转换为股 (对齐 Python: volume * 100)
            bar.amount           = raw.Amount;
            bar.up               = static_cast<int>(raw.UpCount);
            bar.down             = static_cast<int>(raw.DownCount);
            bar.timestamp        = raw.DateTime;
            bar.adjustment_count = 0;
            result.push_back(std::move(bar));
        }

        spdlog::debug("[kline_raw] fetch_kline_raw_from_std: {} bars for {}",
                      result.size(), inst.symbol());
        return result;
    } catch (const std::exception& e) {
        spdlog::error("[kline_raw] fetch_kline_raw_from_std error for {}: {}",
                      inst.symbol(), e.what());
        return {};
    }
}

/// fetch_kline_raw_from_ext — 从扩展行情获取原始K线, 转换为 domain Bar (港股/美股等)
/// 对应 Python kline_raw.py fetch_kline_raw_from_ext:
///   with get_ext_conn() as conn:
///       bars = InstrumentBars(kline_type.value, inst.ext_market, ticker=code.upper(), start=start, count=count)
///       protocol.transact_message_sync(conn, bars)
///       return bars.reply  # bars.reply 已经是 List[Bar]
static std::vector<schema::Bar> fetch_kline_raw_from_ext(
        const meta::Instrument& inst, int start, int count, u16 category) {
    try {
        auto conn = get_ext_conn();
        if (!conn) {
            spdlog::warn("[kline_raw] fetch_kline_raw_from_ext: no ext connection for {}", inst.symbol());
            return {};
        }

        std::string ticker;
        if (!inst.alias_ticker.empty()) {
            ticker = inst.alias_ticker;
        } else {
            ticker = inst.ticker;
        }
        // 对齐 Python: ticker=code.upper()
        for (auto& c : ticker) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));

        InstrumentBars bars(
            static_cast<u8>(inst.ext_market),
            ticker,
            category,
            static_cast<u32>(start),
            static_cast<u16>(count)
        );

        transact_message_sync(conn->socket(), bars);

        spdlog::debug("[kline_raw] fetch_kline_raw_from_ext: {} bars for {}",
                      bars.reply.size(), inst.symbol());
        return std::move(bars.reply);
    } catch (const std::exception& e) {
        spdlog::error("[kline_raw] fetch_kline_raw_from_ext error for {}: {}",
                      inst.symbol(), e.what());
        return {};
    }
}

std::vector<schema::Bar> fetch_kline_raw(
        const meta::Instrument& inst, int start, int count, u16 category) {
    if (exchange_is_std_quote(inst.exchange)) {
        return fetch_kline_raw_from_std(inst, start, count, category);
    } else if (exchange_is_ext_quote(inst.exchange)) {
        return fetch_kline_raw_from_ext(inst, start, count, category);
    }
    return {};
}

// =============================
// DataKLineRaw — 未复权日K线数据适配器
// 对应 Python class DataKLineRaw(DataAdapter)
// =============================

void DataKLineRaw::Print(const meta::Instrument& inst, const meta::Timestamp& date) {
    auto filename = get_kline_raw_filename(inst);
    auto klines = read_kline_raw_from_csv(filename);
    if (klines.empty()) {
        fmt::print("\n=== {}: {} ===\n  (no data)\n", Name(), inst.symbol());
        return;
    }
    if (!date.empty()) {
        std::string date_str = date.only_date();
        klines.erase(std::remove_if(klines.begin(), klines.end(),
            [&](auto const& b) { return b.date > date_str; }), klines.end());
    }
    fmt::print("\n=== {}: {} ({} rows) ===\n", Name(), inst.symbol(), klines.size());
    fmt::print("{:<12} {:>8} {:>8} {:>8} {:>8} {:>12} {:>14} {:>4} {:>4}\n",
               "date", "open", "close", "high", "low", "volume", "amount", "up", "dn");
    fmt::print("{:-<82}\n", "");
    size_t head = std::min<size_t>(klines.size(), 10);
    for (size_t i = 0; i < head; ++i) {
        auto const& b = klines[i];
        fmt::print("{:<12} {:>8.2f} {:>8.2f} {:>8.2f} {:>8.2f} {:>12.0f} {:>14.0f} {:>4} {:>4}\n",
                   b.date, b.open, b.close, b.high, b.low,
                   b.volume, b.amount, b.up, b.down);
    }
    if (klines.size() > 20) {
        fmt::print("  ... {} rows omitted ...\n", klines.size() - 20);
        head = std::min<size_t>(10, klines.size());
        for (size_t i = klines.size() - head; i < klines.size(); ++i) {
            auto const& b = klines[i];
            fmt::print("{:<12} {:>8.2f} {:>8.2f} {:>8.2f} {:>8.2f} {:>12.0f} {:>14.0f} {:>4} {:>4}\n",
                       b.date, b.open, b.close, b.high, b.low,
                       b.volume, b.amount, b.up, b.down);
        }
    } else if (klines.size() > 10) {
        for (size_t i = 10; i < klines.size(); ++i) {
            auto const& b = klines[i];
            fmt::print("{:<12} {:>8.2f} {:>8.2f} {:>8.2f} {:>8.2f} {:>12.0f} {:>14.0f} {:>4} {:>4}\n",
                       b.date, b.open, b.close, b.high, b.low,
                       b.volume, b.amount, b.up, b.down);
        }
    }
}

void DataKLineRaw::Update(const meta::Instrument& inst, const meta::Timestamp& date) {
    (void)date;
    auto symbol = inst.symbol();

    // 1. 从本地缓存确定起始日期
    // 对齐 Python: current_start_date = Timestamp.parse(GLOBAL_DEFAULT_START_DATE)
    meta::Timestamp current_start_date = meta::Timestamp::parse(kGlobalDefaultStartDate);
    auto cache_filename = get_kline_raw_filename(inst);

    // 对齐 Python: cache_klines = read_kline_raw_from_csv(cache_filename)
    auto cache_klines = read_kline_raw_from_csv(cache_filename);

    size_t klines_length = cache_klines.size();
    size_t klines_offset_days = kMaxCachedDaysToDrop;

    if (klines_length > 0) {
        if (klines_offset_days > klines_length) {
            klines_offset_days = klines_length;
        }
        // 对齐 Python: kline = cache_klines[klines_length - klines_offset_days]; current_start_date = Timestamp.parse(kline.date)
        auto& kline = cache_klines[klines_length - klines_offset_days];
        current_start_date = meta::Timestamp::parse(kline.date);
    }

    // 2. 确定结束日期
    auto current_end_date = meta::Timestamp::now().pre_market_time();
    spdlog::debug("[DataKLineRaw] [{}]: from {} to {}",
                  symbol, current_start_date.only_date(), current_end_date.only_date());

    // 3. 分页拉取原始K线 — fetch_kline_raw 返回 domain Bar (对齐 Python: reply = fetch_kline_raw(inst, start, count, freq))
    int step = kSecurityBarsPreRequestMax;
    int start = 0;
    std::vector<std::vector<schema::Bar>> batches;
    while (true) {
        int count = step;
        auto reply = fetch_kline_raw(inst, start, count, static_cast<u16>(KLineType::DAILY));
        if (reply.empty()) break;

        // 对齐 Python: last_bar = reply[-1]; last_bar_date = Timestamp.parse(last_bar.date).get_pre_market_time()
        auto& last_bar = reply.back();
        auto last_bar_date = meta::Timestamp::parse(last_bar.date).pre_market_time();

        batches.push_back(std::move(reply));

        if (last_bar_date < current_start_date) break;
        if (reply.size() < static_cast<size_t>(count)) break;
        start += count;
    }

    // 4. 反转页面 (时间升序, 对齐 Python hs.reverse())
    std::reverse(batches.begin(), batches.end());

    // 5. 构建增量K线并过滤日期范围, 转为 BarRaw 缓存格式
    // 对齐 Python: for vec in reversed(hs): for row in vec: filter by date, create BarRaw(...)
    std::vector<BarRaw> incremental_klines;
    for (auto& batch : batches) {
        for (auto& bar : batch) {
            auto date_time = meta::Timestamp::parse(bar.date).pre_market_time();
            if (date_time < current_start_date || date_time > current_end_date) continue;

            // 对齐 Python: kx = BarRaw(date=date_time.only_date(), open=row.open, ..., volume=row.volume * 100, ...)
            BarRaw bx{
                date_time.only_date(),
                bar.open,
                bar.close,
                bar.high,
                bar.low,
                bar.volume * 100,   // 转换为股 (对齐 Python: volume = row.volume * 100)
                bar.amount,
                bar.up,
                bar.down,
                bar.timestamp
            };
            incremental_klines.push_back(std::move(bx));
        }
    }

    if (incremental_klines.empty()) {
        spdlog::debug("[DataKLineRaw] no new data for {}", symbol);
        return;
    }

    // 6. 合并旧缓存和新数据
    // 对齐 Python: klines = []; if klines_length > klines_offset_days: klines.extend(cache_klines[:...]); klines.extend(incremental_klines)
    std::vector<BarRaw> klines;
    if (klines_length > klines_offset_days) {
        klines.insert(klines.end(),
                      cache_klines.begin(),
                      cache_klines.begin() + (klines_length - klines_offset_days));
    }
    klines.insert(klines.end(), incremental_klines.begin(), incremental_klines.end());

    // 7. 保存到缓存文件 (对齐 Python: save_kline_raw(cache_filename, klines))
    save_kline_raw(cache_filename, klines);

    spdlog::info("[DataKLineRaw] updated {} ({} bars) -> {}",
                 symbol, klines.size(), cache_filename);
}

} // namespace quant1x::contrib::data::tdx

File: quant1x/contrib/data/tdx/bar_raw.go
package tdx

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"

	"gitee.com/quant1x/quant1x/quant1x/data/exchange"
	"github.com/quant1x/quant1x/quant1x/config"
	"github.com/quant1x/quant1x/quant1x/contrib/data/tdx/level1/std"
	"github.com/quant1x/quant1x/quant1x/data"
	"github.com/quant1x/quant1x/quant1x/data/schema"
	"github.com/quant1x/quant1x/quant1x/encoding"
	logger "github.com/quant1x/quant1x/quant1x/log"
)

// DataBarRaw 实现了与 C++ DataBarRaw 类似的缓存适配器更新逻辑
type DataBarRaw struct{}

// 实现 DataAdapter 的 Schema 方法
func (d *DataBarRaw) Kind() data.Kind { return data.BaseRawDailyKLine }
func (d *DataBarRaw) Owner() string   { return data.DefaultDataProvider }
func (d *DataBarRaw) Key() string     { return "day_raw" }
func (d *DataBarRaw) Name() string    { return "日K线RAW" }
func (d *DataBarRaw) Usage() string   { return "日K线RAW" }

// Print 实现 DataAdapter.Print(可变参数日期)
func (d *DataBarRaw) Print(code data.InstrumentInfo, dates ...exchange.Timestamp) {
	_ = code
	_ = dates
}

// Update 对应 C++ DataBarRaw::Update 的行为: 读取本地缓存, 确定时间窗口, 分页拉取 level1 数据,
// 反转与合并结果, 并写回缓存文件.
func (d *DataBarRaw) Update(code data.InstrumentInfo, _date data.Timestamp) {
	// 1. 确定缓存文件并读取本地缓存
	cacheFilename := config.GetBarFilename(code.Symbol(), false)
	var cacheBars []BarRaw
	encoding.CsvToSlices(cacheFilename, &cacheBars)

	barsLength := len(cacheBars)
	barsOffsetDays := schema.MaxCachedDaysToDropOnIncrementalUpdate

	// 默认起始日期(使用 datasets.MarketFirstDate, 与 C++ 的 market_first_date 等价)
	currentStartDate := data.GetFirstMarketDate(code.Exchange)
	if barsLength > 0 {
		if barsOffsetDays > barsLength {
			barsOffsetDays = barsLength
		}
		bar := cacheBars[barsLength-barsOffsetDays]
		if ts, err := data.ParseTimestamp(bar.Date); err == nil {
			currentStartDate = ts
		}
	}

	// 2. 确定结束日期
	currentEndDate := data.NowTimestamp().PreMarketTime()
	logger.Debugf("[dataset::BarRaw] [%s]: from %s to %s", code, currentStartDate.OnlyDate(), currentEndDate.OnlyDate())

	step := uint16(std.SecurityBarsMax)
	start := uint16(0)
	hs := make([][]std.SecurityBar, 0)
	elementCount := 0

	for {
		count := step
		reply, err := tdxFetchRawSecurityBars(code, std.FreqDaily, start, count)
		if err != nil || len(reply) == 0 {
			break
		}

		elementCount += len(reply)
		hs = append(hs, reply)

		lastBar := reply[len(reply)-1]
		lastBarDate := data.PreMarketTimestamp(lastBar.Year, lastBar.Month, lastBar.Day)

		if lastBarDate.Less(currentStartDate) {
			break
		}

		if len(reply) < int(count) {
			break
		}

		start += uint16(count)
	}

	// 反转切片以获得正确的顺序
	for i, j := 0, len(hs)-1; i < j; i, j = i+1, j-1 {
		hs[i], hs[j] = hs[j], hs[i]
	}

	incrementalBars := make([]BarRaw, 0)

	for _, vec := range hs {
		for _, row := range vec {
			dateTime := data.PreMarketTimestamp(row.Year, row.Month, row.Day)

			if dateTime.Less(currentStartDate) || dateTime.Greater(currentEndDate) {
				continue
			}

			bx := BarRaw{
				Date:     dateTime.OnlyDate(),
				Open:     row.Open,
				Close:    row.Close,
				High:     row.High,
				Low:      row.Low,
				Volume:   row.Vol * 100, // Convert to shares
				Amount:   row.Amount,
				Up:       int(row.UpCount),
				Down:     int(row.DownCount),
				Datetime: row.DateTime,
			}
			incrementalBars = append(incrementalBars, bx)
		}
	}

	// 7. 合并
	bars := make([]BarRaw, 0)
	if barsLength > barsOffsetDays {
		bars = append(bars, cacheBars[:barsLength-barsOffsetDays]...)
	}
	bars = append(bars, incrementalBars...)

	// 9. 保存
	if err := saveBarRaw(cacheFilename, bars); err != nil {
		logger.Errorf("[dataset::BarRaw] save error: %v", err)
	}
}

func init() {
	// 注册DataBarRaw插件
	if err := data.Register(&DataBarRaw{}); err != nil {
		logger.Errorf("[dataset::BarRaw] failed to register plugin: %v", err)
	}
}

// BarRaw 对应 C++ 中的 data::BarRaw, 用于表示原始日线数据.
type BarRaw struct {
	Date     string  `name:"日期" csv:"date"`        // 日期 YYYY-MM-DD
	Open     float64 `name:"开盘价" csv:"open"`       // 开盘价
	Close    float64 `name:"收盘价" csv:"close"`      // 收盘价
	High     float64 `name:"最高价" csv:"high"`       // 最高价
	Low      float64 `name:"最低价" csv:"low"`        // 最低价
	Volume   float64 `name:"成交量(股)" csv:"volume"`  // 成交量(股)
	Amount   float64 `name:"成交额(元)" csv:"amount"`  // 成交额(元)
	Up       int     `name:"上涨家数 / 外盘" csv:"up"`   // 上涨家数 / 外盘
	Down     int     `name:"下跌家数 / 内盘" csv:"down"` // 下跌家数 / 内盘
	Datetime string  `name:"时间" csv:"datetime"`    // 时间
}

//const maxBarRawLookbackDays = 1

// saveBarRaw 将原始 K 线写入 CSV 文件(包含表头).
func saveBarRaw(filename string, values []BarRaw) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()

	header := []string{"date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime"}
	if err := w.Write(header); err != nil {
		return err
	}
	for _, r := range values {
		row := []string{
			r.Date,
			fmt.Sprintf("%g", r.Open),
			fmt.Sprintf("%g", r.Close),
			fmt.Sprintf("%g", r.High),
			fmt.Sprintf("%g", r.Low),
			fmt.Sprintf("%g", r.Volume),
			fmt.Sprintf("%g", r.Amount),
			fmt.Sprintf("%d", r.Up),
			fmt.Sprintf("%d", r.Down),
			r.Datetime,
		}
		if err := w.Write(row); err != nil {
			return err
		}
	}
	return nil
}

// ReadBarRawFromCSV 从 CSV 文件中读取 BarRaw 条目；解析错误会返回错误.
func ReadBarRawFromCSV(filename string) ([]BarRaw, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	r := csv.NewReader(f)
	rows, err := r.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}
	// 期望第一行为表头
	out := make([]BarRaw, 0, len(rows)-1)
	for i := 1; i < len(rows); i++ {
		rec := rows[i]
		if len(rec) < 10 {
			continue
		}
		open, _ := strconv.ParseFloat(rec[1], 64)
		closev, _ := strconv.ParseFloat(rec[2], 64)
		high, _ := strconv.ParseFloat(rec[3], 64)
		low, _ := strconv.ParseFloat(rec[4], 64)
		vol, _ := strconv.ParseFloat(rec[5], 64)
		amt, _ := strconv.ParseFloat(rec[6], 64)
		up, _ := strconv.Atoi(rec[7])
		down, _ := strconv.Atoi(rec[8])
		datetime := rec[9]
		k := BarRaw{
			Date:     rec[0],
			Open:     open,
			Close:    closev,
			High:     high,
			Low:      low,
			Volume:   vol,
			Amount:   amt,
			Up:       up,
			Down:     down,
			Datetime: datetime,
		}
		out = append(out, k)
	}
	return out, nil
}

// LoadBarRaw 从缓存文件加载原始K线数据.
func LoadBarRaw(code string) ([]BarRaw, error) {
	filename := config.GetBarFilename(code, false)
	logger.Debugf("[data::BarRaw] bar file: %s", filename)
	return ReadBarRawFromCSV(filename)
}
